### DISCUSSION TITLE
Is inconsistency between the identity map and queries expected by default for SQLite?

### DISCUSSION THREAD

**Discussion by ntc2** [sqlalchemy/sqlalchemy#6622] [Usage Questions] [answered]
I eventually found the SQLAlchemy docs section where it explains that [transactions are broken by default for SQLite](https://docs.sqlalchemy.org/en/14/dialects/sqlite.html?highlight=sqlite#pysqlite-serializable), because the corresponding Python standard library backend is broken, and using the suggestions there I was able to fix my problem. But I'm still baffled by the default behavior I'm observing, and wanted to make sure this is really expected, since it was quite tricky to debug my original problem "in the wild".

Here's a simplified version of my problem. I have a program `query.py`:

```python
import time
from models import Ball, session

time.sleep(1)
r = session.query(Ball).filter(Ball.color=='red').first()
print(f'Red ball color is {r.color}')
time.sleep(2)
b = session.query(Ball).filter(Ball.color=='blue').first()
print(f'Blue ball color is {b.color}')
print(f'Red ball id is {r.id}, blue ball id is {b.id}')
```

When I run `query.py` at the same time as `modify.py` (included below), I get the following output:

```console
$ python modify.py &! python query.py
Red ball color is red                                                                                                                                                   
Blue ball color is red                                                                                                                                                  
Red ball id is 1, blue ball id is 1                                                                                                                  
```

The problem is that the blue ball is red!

Here is the content of `models.py`:

```python
import sqlalchemy as sa
import sqlalchemy.orm as sao
import sqlalchemy.ext.declarative as saed

Base = saed.declarative_base()

class Ball(Base):
    __tablename__ = 'ball'
    id = sa.Column(sa.Integer, primary_key=True)
    color = sa.Column(sa.String)

engine = sa.create_engine('sqlite:///test.db')
Base.metadata.create_all(engine)
session = sao.Session(engine)
```

And here is `modify.py`:

```python
import time
from models import Ball, session

session.query(Ball).delete()
b = Ball(color='red')
session.add(b)
session.commit()
time.sleep(2)
b.color = 'blue'
session.add(b)
session.commit()
```

I find it very strange that I get an inconsistency here between my DB query (that sees the latest DB state) and the object returned via the SQLAlchemy identiy map for my DB query (which is stale, reflecting the DB state the first time the row in question was read). Is this expected?

I would have been happy if the blue ball were blue -- i.e. if the DB query and object it returned agreed -- or if the blue ball query returned `None` -- i.e. if the concurrent DB modification was not visible in the query transaction. But I seem to be stuck in the middle by default, and found this incredibly confusing when I ran into it in a real project. I've since figured out how to get both of these "reasonable" behaviors, by using `session.expire_all()` before all my queries, or by changing SQLite transactions to be "SERIALIZABLE" by following [the docs](https://docs.sqlalchemy.org/en/14/dialects/sqlite.html?highlight=sqlite#pysqlite-serializable), respectively.

I'm using SQLAlchemy versions 1.4.15, Python version 3.9.1, and SQLite version 3.22.0. [I tried asking about this on SO](https://stackoverflow.com/questions/67858384/sqlalchemy-how-do-i-avoid-this-inconsistency-between-orm-cache-and-db/67890633#67890633) but no one responded.

**Comment by zzzeek** [comment, marked as answer]
hi -

we have an in-depth explaination of this behavior at the FAQ entry https://docs.sqlalchemy.org/en/14/faq/sessions.html#i-m-re-loading-data-with-my-session-but-it-isn-t-seeing-changes-that-i-committed-elsewhere .    Go through that and come back here for clarifications.

**Reply by ntc2** [reply]
Thanks! That page helpfully explains that inconsistency between queries and the objects they return is expected when the returned objects have already been read in the current transaction, but the isolation level is less than "Repeatable Reads". Then I guess it remains for me to understand what isolation behavior I'm getting, and understand my transaction scopes.

Regarding transaction scope, I was very surprised when I read that the first suggested solution is to use short transactions, since I thought that's what I was already doing! My real application is structured as two processes, a REST API that submits jobs to the DB, and a worker daemon that repeatedly queries the DB for new jobs and completes them. Both processes do their work and commit the results immediately. However, it turns out my daemon transactions are actually long lived in some cases, because *after* committing a daemon transaction I write a log message describing what action was completed: the log message includes printing an ORM object, and printing an ORM object after `session.commit()` results in an implicit `SELECT` to refresh the fields of the ORM object, which in turn starts a new transaction. In case that's not clear, the daemon is roughly like this:

```python
while True:
    time.sleep(1)
    job = session.query(Job).filter(Job.status=='new').first()
    if job is not None:
        <process job>
        job.status = 'done'
        session.commit()
        # Implicit transaction starts here and reloads job!
        log.info(f'Processed {job}')
```

and the problem arises when the most recently completed job gets reset to the "new" status without any intervening jobs being processed. I found this very subtle and it took me a few hours to track it down, but at least i probably won't make this mistake again going forward ...

In terms of the example in my original question, this is similar to trying to fix the problem by inserting `session.commit()` or `session.expire_all()` after printing the red ball, but then printing the red ball again. I.e.

```python
import time
from models import Ball, session

time.sleep(1)
r = session.query(Ball).filter(Ball.color=='red').first()
print(f'Red ball color is {r.color}')
session.commit()
print(f'Done with {r.color} colored ball')
time.sleep(2)
b = session.query(Ball).filter(Ball.color=='blue').first()
print(f'Blue ball color is {b.color}')
print(f'Red ball id is {r.id}, blue ball id is {b.id}')
```

produces the output

```
$ python modify.py &! python query.py
Red ball color is red
Done with red colored ball
Blue ball color is red
Red ball id is 1, blue ball id is 1
```

Similar to the `session.comit()` there not helping, I get the same result if I add `session.expire_all()`. But `session.expunge_all()` and `session.close()` both get the correct result, I guess because they sever the connection between the ORM object and the DB (but I'm still not completely clear on this).

Regarding isolation, it seems that SQLite + SQLAlchemy defaults to something between "Read Committed" and "Serializable": my experiments indicate that when a read-only transaction overlaps with a write transaction, the read-only transaction sees the writes as they're committed. But when two write transactions overlap, the second blocks until the first completes.
